### ISSUE TITLE
Render Editor viewport gizmos and UI on a render pass

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#2138]
### Describe the project you are working on
This applies to any 3D project I've worked on so far

### Describe the problem or limitation you are having in your project
I can't see the viewport UI/gizmos clearly because the WorldEnvironment post processing is applied to the viewport UI/Gizmos as well. It drives me crazy.
![image](https://user-images.githubusercontent.com/47016402/104838516-9fea6780-58bb-11eb-89ea-88788846bd5e.png)

This has been discussed before, but those issues have been closed. Why, I really don't understand.
https://github.com/godotengine/godot/issues/8140
https://github.com/godotengine/godot/issues/8485
["There is really no fixing to this, you can disable environment now in individual viewports, so closing this."](https://github.com/godotengine/godot/issues/8485#issuecomment-307987209)
Disabling the WorldEnvironment is unfortunately not really an option because the WE is so substantial and intertwined to any 3D scene lighting. As level artist, environment artist, art director game designer and concept artist I cannot evaluate what I am doing/what needs to be done without WE on because the scene would look completely differently.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
It would help me to see all the little things I have to constantly click and drag.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
The viewport gizmos should look like the axis in the top right corner in the above screenshot, unaffected by the light and post processing of the game world. Unfortunately I'm not a 3D graphics programmer, so I don't know anything about the implementation.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Would be used by anyone working in 3D.

### Is there a reason why this should be core and not an add-on in the asset library?
It's a basic usability necessity.

**Comment by Calinou** [comment]
If gizmos are drawn on a separate render pass, then they won't be able to be occluded by the scene's geometry anymore.

**Comment by golddotasksquestions** [comment]
`A simpler way to fix this would be to have a toggle or editor setting to always disable glow, auto exposure and depth of field while in the editor.
`
I really disagree strongly. This might be simple to implement, but it's definitely not solution. Glow, auto-exposure, bloom and DOF change the look of a scene drastically. Much more so when your project is stylized. Godot rendering is not great at realism, so people use it for stylized art first and foremost.

When I place an asset, create the detail of the scene, when I paint the splatmap, when I come up with the atmosphere, when I compose the layout of the environment ... I need to be able to see what the end result will look like **as I am doing it.**

Godot's post-processing UX already sucks tremendously, precisely because the Editor scene is always looks slightly different from the running scene and the intertwined WE setting result in super tight tolerances where 0.03 looks ok but 0.01 or 0.06 look terrible  ... resulting in having to have to resort to editing those numbers numerically because the slider won't work in those fine details floating point position any more ... which is again: guesswork. Please let's not make this ordeal worse by further facilitating guesswork rather than fixing the issue properly.

**Comment by Jummit** [comment]
While I don't like your attitude to Godot's capabilities, I agree that at some point this problem has to be addressed. I don't think that it would be a huge problem if the gizmos wouldn't be occluded by geometry, I actually think this is how it works in UE4.

**Comment by Calinou** [comment]
> I don't think that it would be a huge problem if the gizmos wouldn't be occluded by geometry, I actually think this is how it works in UE4.

It would look pretty strange for the grid and origin lines.

**Comment by Zireael07** [comment]
Grid/origin lines could stay as-is IMHO, the main problem is gizmos being affected by glow/what have you.

**Comment by golddotasksquestions** [comment]
@Jummit I'm using Godot exclusively and daily now since more than two years, full time. If I would not believe in Godots capabilities, I would not have spend all this time learning, using, and celebrating the community effort that this engine is. 
But just because I love this engine, the people making, improving and using it, does not mean I've become desensitized to it's (sometimes glaring) flaws.
Everyone has different usecases, I can only report on mine. 
I'm not criticizing the contributors who worked on an issue I'm reporting on (If that's the why you mention "attitude"). My sole intention is to bring attention to what I consider a sizeable stumbling block on Godot's path to glory as well as an impediment to my own daily work.

**Comment by Jummit** [comment]
@golddotasksquestions I agree with most of your criticism, just not with how you express it. Anyway, I'm would like it if we could stay on-topic.

**Comment by alexfreyre** [comment]
I agree with this proposal because with post-process effects it is impossible to see the gizmos, a great example is desaturate completely the scene, the gizmo is almost indistinguishable.

**Comment by dueddel** [comment]
I know it's not really necessary to confirm that issue with more screenshots, but I can't help. I think the problem is even worse with **DoF Blur**.

These are some environment settings
![Bildschirmfoto 2021-04-23 um 09 33 22](https://user-images.githubusercontent.com/436636/115841994-abe74780-a41d-11eb-9a54-3b88ee7a6887.png)

This is what a simple Spatial node's gizmo looks like with above environment settings:
![Bildschirmfoto 2021-04-23 um 09 37 19](https://user-images.githubusercontent.com/436636/115841998-ac7fde00-a41d-11eb-93a9-a6291c5cabde.png)

Blurred gizmo doesn't come in very handy when you need some overview of your scene (e.g. if you wanted to arrange some level assets):
![dof-blur](https://user-images.githubusercontent.com/436636/115843181-d980c080-a41e-11eb-98bc-1b793cde5b60.gif)

However, I too would love to see a fix for that.

**Comment by Calinou** [comment]
> I know it's not really necessary to confirm that issue with more screenshots, but I can't help. I think the problem is even worse with DoF Blur.

This has been solved in the `master` branch by moving DoF Blur settings to a separate CameraEffects resource, which is disabled in the editor unless you're currently previewing a camera node with a CameraEffects resource.

**Comment by lentsius-bark** [comment]
@Calinou I'd like to chip in to say that adding a toggle that'd make blur, bloom and colour corrections game only, that way we could at least set things up, make sure they work and then toggle it.

I see DoF blur referenced a lot but let's not omit the fact that bloom, colour adjustments and colours can also make working with gizmos in 3D awry.

**Comment by Jaku-San** [comment]
Would be real nice QOL to be able to use gizmo when in overdraw mode, and other cases of course

**Comment by paddy-exe** [comment]
Chiming in here since the Colors of the Gizmos shouldn't be dependant on the Sky Background either:
 
![image](https://user-images.githubusercontent.com/38077837/215340003-f24d07da-b0a8-4a88-bb95-231528cbe719.png)

**Comment by mieldepoche** [comment]
Also worth mentionning that gizmos can be altered, wrongly colored, partially or even completely hidden in debug display modes:


[simplescreenrecorder-2023-09-26_05.16.48.webm](https://github.com/godotengine/godot-proposals/assets/71937306/fc810ef5-a973-49bd-9d4b-137cf3288882)

**Comment by atirut-w** [comment]
At this point, I think it's proven that this should have at least medium priority, considering that gizmos are not supposed to be affected by post processing settings in the first place and that other engines seem to also draw gizmos in a separate pass.

Regarding gizmos occlusion, I stand by my suggestion to carry the depth map in the first pass into the next, unless it's impossible at the moment.

**Comment by Calinou** [comment]
Now that we have https://github.com/godotengine/godot/pull/81286, it should be possible to make all 3D gizmos ignore fog.

**Comment by Zireael07** [comment]
@Calinou Fog is only one of the problems. Blur and DOF are the other common offenders

**Comment by Nagasadri** [comment]
I'm just here to report volumetric fog is also affected, making the UI disappear into the fog with distance.
![image](https://github.com/godotengine/godot-proposals/assets/79308545/fad82f93-45a8-4496-a8e1-6a1a6be4e68b)
![image](https://github.com/godotengine/godot-proposals/assets/79308545/4ede0388-27cc-40f6-a623-8145da113b4c)

I vote in favor of the separate pass, if possible with occlusion.
I understand it wouldn't be a small undertaking, but I think it would be worth it for the non-fog effects mentioned earlier some time down the line.

> Now that we have [godotengine/godot#81286](https://github.com/godotengine/godot/pull/81286), it should be possible to make all 3D gizmos ignore fog by adding the **Disable Fog** flag on their materials.

That sounds great, but how does one go about modifying the gizmo material to do that ?
It might be worth considering flipping that flag on by default. I can't think of a use case where I'd want my gizmos to receive fog.

**Comment by marcinn1** [comment]
> Now that we have [godotengine/godot#81286](https://github.com/godotengine/godot/pull/81286), it should be possible to make all 3D gizmos ignore fog by adding the **Disable Fog** flag on their materials.

Great!! How about glow?

**Comment by Calinou** [comment]
> Great!! How about glow?

This also requires drawing to a separate pass, as there is no way to exclude specific objects from glow.

**Comment by Xorblax** [comment]
Being able to change the gizmo material would be really great because all I need to do is make light gizmos ignore depth completely and my problem would be solved.

**Comment by atirut-w** [comment]
Considering that Godot now have the ability to use hooks to insert post processing, inserting a render pass for gizmos might be a little easier now?

**Comment by darkhog** [comment]
From: https://github.com/godotengine/godot-proposals/discussions/9773#discussioncomment-9484350

Currently the gizmos seem to be rendered in the same pass as the rest of the scene, meaning they're susceptible to several issues:

\- Getting discolored due to the weird World Environment settings:  
!\[obraz\](https://github.com/godotengine/godot-proposals/assets/1531474/edacd57c-94e5-4a68-a984-ff2dd1e7116b)  
\- Not being rendered properly with certain shaders (notice how the arrow showing the direction of the directional light gets cut out when the water is rendered, as if the arrow gets rendered under the water shader:  
!\[obraz\](https://github.com/godotengine/godot-proposals/assets/1531474/ffbc8e12-8fde-41a4-9026-7e7f50cbed2b)

Having gizmos being rendered in a separate pass after everything else gets rendered would fix both of these issues as the gizmo lighting settings could get reset so that gizmos always have proper colors regardless of WorldEnvironment settings used and (due to it being a separate pass) the gizmos would always be rendered on top of everything else regardless of what shaders the developer uses.

**Comment by Calinou** [comment]
> Not being rendered properly with certain shaders (notice how the arrow showing the direction of the directional light gets cut out when the water is rendered, as if the arrow gets rendered under the water shader:

That issue can be fixed by using opaque rendering for line gizmos, although some gizmos are transparent so it'd need to be toggled on a per-line basis.

Now that https://github.com/godotengine/godot/pull/75303 has been merged, a lot of gizmos will only be displayed as fully opaque (or not displayed at all), so we can probably change the transparency mode for a lot of line gizmos.

**Comment by Rinocrosser** [comment]
@Calinou It seems absolutely unclear right now regarding when these elements will be placed in a different pass? It isn't fixed in 4.2 and 4.3 (some issues are fixed) and the proposal isn't marked as 4.x at all. Can you specify at least briefly?

**Comment by Calinou** [comment]
> @Calinou It seems absolutely unclear right now regarding when these elements will be placed in a different pass? It isn't fixed in 4.2 and 4.3 (some issues are fixed) and the proposal isn't marked as 4.x at all. Can you specify at least briefly?

There is no set milestone for this proposal, as nobody has started working on implementing this.

Also, there are ways to resolve this issue without using a different pass:

- For exposure, compensate the scene exposure in the gizmo colors.
- For depth of field, disable depth of field when a node is selected by default.

**Comment by jcostello** [comment]
DoF is not the only problem affecting gizmos. Glow also affects, color grading, etc.  A different pass will be ideal so no even a custom shader will affect the gizmos

**Comment by Rinocrosser** [comment]
> > @Calinou It seems absolutely unclear right now regarding when these elements will be placed in a different pass? It isn't fixed in 4.2 and 4.3 (some issues are fixed) and the proposal isn't marked as 4.x at all. Can you specify at least briefly?
> 
> There is no set milestone for this proposal, as nobody has started working on implementing this.
> 
> Also, there are ways to resolve this issue without using a different pass:
> 
> * For exposure and adjustments (color correction), compensate the scene exposure and adjustments in the gizmo colors.
> * For depth of field, disable depth of field when a node is selected by default (unless a camera is being previewed).
> 
> For depth of field in particular, you can disable it in the editor only by putting the CameraEffects resource in the Camera3D node instead of WorldEnvironment. This is the recommended approach in the documentation. When doing so, depth of field remains visible when previewing the camera.

All of the above is like moving to another room when there's a fire in your room: not a solution, but a cover-up for the issue (and not even a usable one). The only solution is to move it into another pass.

**Comment by atirut-w** [comment]
Yeah, why are you guys so weirdly against putting gizmos on a separate render pass? Worst case scenario you literally just add a second camera that follows the main one and only renders the gizmos layer and composite it onto the main viewport. I've done a similar thing in the past in Unity to implement moon without editing the sky shader, *it won't kill you*.

**Comment by darkhog** [comment]
Yep, rendering them in a separate pass is the best way. It won't be too draining on the GPU resources (as those are just gizmos) and won't affect exported games anyway as gizmos are only visible in the editor.

**Comment by clayjohn** [comment]
I'd like to add a little bit of clarity here. 

Yes, the suggestion to add gizmos to a separate render pass is good. However, before we do that, we need to add the ability to create and use separate render passes in the same viewport. Currently, Godot does only one render pass per viewport, so our options are to:
A) render gizmos in a separate Viewport (and lose access to to depth information) or
B) render the gizmos in the same viewport and workaround the problems.

Neither one of those options are great and frankly, both are bad enough that its not worth our time to work on them.

The only good option is to implement the ability to have multiple render passes and _then_ move editor gizmos into a separate render pass. I would love if we had the ability to create multiple render passes in a single viewport already, but the Rendering team is very short on volunteers willing to take on big projects like that, so there has been no visible progress. 

As always, the way it works with open source is that there is no progress until someone is willing to do the work.

**Comment by Servail** [comment]
I've just stumbled across this problem trying to do HDR lighting - gizmos are dimming out with auto exposure, and getting blurred with DoF. I can bear with that, but gizmos definitely should be drawn over the scene, and fade/shrink (if specified) based on camera distance. "Spatial gizmos" must be a preference choice. And yes, you can't just disable auto exposure when working with HDR lighting, at most you can use fixed exposure but it will be the same high working value. Though dimming could be potentially fixed by multiplying gizmos color according to exposure, we need the real solution (that works for custom effects too), no matter dedicated Viewport layer or separate pass.

**Comment by darkhog** [comment]
> If gizmos are drawn on a separate render pass, then they won't be able to be occluded by the scene's geometry anymore. I'm not sure if this can be fixed.

That doesn't need fixing and is in fact preferred over having gizmos for a house being occluded by some random fence.

**Comment by Calinou** [comment]
> That doesn't need fixing and is in fact preferred over having gizmos for a house being occluded by some random fence.

See my comment in https://github.com/godotengine/godot/pull/115821#issuecomment-3842029385:

> That said, rendering gizmos in a separate viewport will break any depth-related effects, causing them to always render in front of everything. This is bound to look confusing in large scenes with lots of overdraw. Depth perception is often pretty important when you're trying to place things in the 3D world.

Depth effects don't have to be "all or nothing". You can render something with lower opacity when it's occluded by rendering it twice (once opaque with depth test, once translucent with no depth test). This is currently used for the 3D selection box:

<img width="796" height="612" alt="Image" src="https://github.com/user-attachments/assets/a0e38525-ba17-44a5-98ee-2d2cc8635f2c" />

In any case, the editor already allows you to make any 3D gizmo visible through solid surfaces:

<img width="692" height="353" alt="Image" src="https://github.com/user-attachments/assets/2a3e139b-553b-41ed-afb8-e0c76cace849" />

Click twice to switch to x-ray status (partially filled eye icon).

**Comment by darkhog** [comment]
Those are icons, not gizmos (even if Godot calls them that). The gizmos are the things you move/rotate/scale objects with.

**Comment by marcinn** [comment]
In some cases X-ray or solid gizmos looks pretty similar:

<img width="879" height="578" alt="Image" src="https://github.com/user-attachments/assets/03640739-6271-428b-8ef9-3b89ac5d0b8b" />

My patch breaks the transparency, but as an user I like this more:
<img width="1026" height="669" alt="Image" src="https://github.com/user-attachments/assets/76454461-77ea-4a7d-b880-87e83b08804a" />

I didn't know about x-ray gizmos... :/ 
I believe we can find some solution. Will try something else.

**Comment by TokageItLab** [comment]
> That doesn't need fixing and is in fact preferred over having gizmos for a house being occluded by some random fence.

Some gizmos, such as Skeleton3D, utilize a hack to achieve this. They draw the mesh on a plane very close to the camera and crossed to the camera's forward axis, compressing its scale along that axis to a very small value via the vertex shader by moving position. If this is ever fixed, note that this shader hack will need to be removed.

```cpp
	selected_sh->set_code(R"(
// Skeleton 3D gizmo bones shader.

shader_type spatial;
render_mode unshaded, shadows_disabled;

void vertex() {
	if (!OUTPUT_IS_SRGB) {
		COLOR.rgb = mix(pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)), COLOR.rgb * (1.0 / 12.92), lessThan(COLOR.rgb,vec3(0.04045)));
	}
	VERTEX = VERTEX;
	POSITION = PROJECTION_MATRIX * VIEW_MATRIX * MODEL_MATRIX * vec4(VERTEX.xyz, 1.0);
	POSITION.z = mix(POSITION.z, POSITION.w, 0.998);
}

void fragment() {
	ALBEDO = COLOR.rgb;
	ALPHA = COLOR.a;
}
)");
```
